precompiles/relay-data-verifier/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright 2024 Moonbeam Foundation.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Precompile for verifying relay entries against a relay block number.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::marker::PhantomData;
+use cumulus_primitives_core::relay_chain::BlockNumber as RelayBlockNumber;
+use fp_evm::{PrecompileFailure, PrecompileHandle};
+use frame_support::{ensure, traits::ConstU32};
+use precompile_utils::prelude::*;
+use sp_core::H256;
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub mod proof;
+mod weights;
+use proof::{ProofError, StorageProofChecker};
+
+pub const CALL_DATA_LIMIT: u32 = 2u32.pow(16);
+pub const ARRAY_LIMIT: u32 = 2048;
+pub const KEY_LENGTH_LIMIT: u32 = 256;
+
+pub type GetCallDataLimit = ConstU32<CALL_DATA_LIMIT>;
+pub type GetKeyLengthLimit = ConstU32<KEY_LENGTH_LIMIT>;
+pub type GetArrayLimit = ConstU32<ARRAY_LIMIT>;
+
+pub type RawKey = BoundedBytes<GetKeyLengthLimit>;
+
+/// Relay Data Verifier precompile.
+pub struct RelayDataVerifierPrecompile<Runtime>(PhantomData<Runtime>);
+
+#[precompile_utils::precompile]
+impl<Runtime> RelayDataVerifierPrecompile<Runtime>
+where
+	Runtime: frame_system::Config + pallet_relay_storage_roots::Config + pallet_evm::Config,
+{
+	/// Verify the storage entry using the provided relay block number and proof. Return the value
+	/// of the storage entry if the proof is valid and the entry exists.
+	#[precompile::public("verifyEntry(uint32,(bytes32,bytes[]),bytes)")]
+	#[precompile::public("verify_entry(uint32,(bytes32,bytes[]),bytes)")]
+	fn verify_entry(
+		handle: &mut impl PrecompileHandle,
+		relay_block_number: RelayBlockNumber,
+		proof: ReadProof,
+		key: RawKey,
+	) -> EvmResult<UnboundedBytes> {
+		let weight = weights::WeightInfo::<Runtime>::verify_entry(proof.proof.len() as u32);
```

Comment:
Theses benchamrks account only for the proof verification but not for entries reads, can you account one db read per key?